tests/test_config.py

```diff
@@ -0,0 +1,299 @@
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+import attr
+import pytest
+import sys
+
+from charmcraft.cmdbase import CommandError
+from charmcraft.config import (
+    _BasicPrime,
+    _CharmhubConfig,
+    _check_url,
+    load,
+)
+
+# Decide if we're using a Python 3.5 or older to support a jsonschema detail that uses
+# randomly ordered dictionaries
+is_py35 = (sys.version_info.major, sys.version_info.minor) < (3, 6)
+
+
+@pytest.fixture
+def create_config(tmp_path):
+    """Helper to create the config."""
+    def create_config(text):
+        test_file = tmp_path / "charmcraft.yaml"
+        test_file.write_text(text)
+        return tmp_path
+    return create_config
+
+
+# -- tests for the config loading
+
+def test_load_current_directory(create_config, monkeypatch):
+    """Init the config using charmcraft.yaml in current directory."""
+    tmp_path = create_config("""
+        type: charm
+    """)
+    monkeypatch.chdir(tmp_path)
+    config = load(None)
+    assert config.type == 'charm'
+    assert config.project.dirpath == tmp_path
+
+
+def test_load_specific_directory_ok(create_config):
+    """Init the config using charmcraft.yaml in a specific directory."""
+    tmp_path = create_config("""
+        type: charm
+    """)
+    config = load(tmp_path)
+    assert config.type == 'charm'
+    assert config.project.dirpath == tmp_path
+
+
+def test_load_optional_charmcraft_missing(tmp_path):
+    """Specify a directory where the file is missing."""
+    config = load(tmp_path)
+    assert config is None
+
+
+# -- tests for schema restrictions
+
+@pytest.fixture
+def check_schema_error(tmp_path):
+    """Helper to check the schema error."""
+    def check_schema_error(*expected_msgs):
+        """The real checker.
+
+        Note this compares for multiple messages, as for Python 3.5 we don't have control on
+        which of the verifications it will fail. After 3.5 is dropped this could be changed
+        to receive only one message and compare with equality below, not inclusion.
+        """
+        with pytest.raises(CommandError) as cm:
+            load(tmp_path)
+        assert str(cm.value) in expected_msgs
+    return check_schema_error
+
+
+def test_schema_no_extra_properties(create_config, check_schema_error):
+    """Schema validation, can not add undefined properties."""
+    create_config("""
+        type: bundle
+        whatever: new-stuff
+    """)
+    check_schema_error("Additional properties are not allowed ('whatever' was unexpected)")
+
+
+def test_schema_type_mandatory(create_config, check_schema_error):
+    """Schema validation, type is mandatory."""
+    create_config("""
+        someconfig: None
+    """)
+    if is_py35:
+        check_schema_error(
+            "Bad charmcraft.yaml content; missing fields: type.",
+            "Bad charmcraft.yaml content; the 'type' field must be one of: 'charm', 'bundle'.",
```

Comment:
So do this `the 'type' field must be one of: 'charm', 'bundle'` not surface on 3.6+?

Reply:
No. In 3.6+, if you don't include the `type` field it errors out indicating (consistently, not randomly like in 3.5) that the field is missing. The problem is that in 3.5 it fails sometimes with one error, sometimes with the other one. Note that this is internal to jsonschema.